components/constellation/constellation.rs

```diff
@@ -2125,6 +2118,67 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
         }
     }
 
+    fn notify_history_changed(&self, pipeline_id: PipelineId) {
+        // Send a flat projection of the history.
+        // The final vector is a concatenation of the LoadData of the past entries,
+        // the current entry and the future entries.
+        // LoadData of inner frames are ignored and replaced with the LoadData of the parent.
+
+        let top_level_frame_id = self.get_top_level_frame_for_pipeline(pipeline_id);
+
+        // Ignore LoadData of non-top-level frames.
+        let keep_load_data_if_top_frame = |state: &FrameState| {
+            match state.pipeline_id {
+                None => Some(state.load_data.clone()),
+                Some(pipeline_id) => {
+                    match self.pipelines.get(&pipeline_id) {
+                        None => Some(state.load_data.clone()),
+                        Some(pipeline) => match pipeline.parent_info {
+                            None => Some(state.load_data.clone()),
+                            Some(_) => None,
+                        }
+                    }
+                }
+            }
+        };
+
+        // If LoadData was ignored, use the LoadData of the previous FrameState, which
+        // is the LoadData of the parent frame.
+        let resolve_load_data = |previous_load_data: &mut LoadData, load_data| {
+            let load_data = match load_data {
+                None => previous_load_data.clone(),
+                Some(load_data) => load_data,
+            };
+            *previous_load_data = load_data.clone();
+            Some(load_data)
+        };
+
+        let current_load_data = match self.frames.get(&top_level_frame_id) {
+            Some(frame) => frame.load_data.clone(),
+            None => return warn!("notify_history_changed error after top-level frame closed."),
+        };
+
+        let mut entries: Vec<LoadData> = self.joint_session_past(top_level_frame_id)
+            .map(&keep_load_data_if_top_frame)
+            .scan(current_load_data.clone(), &resolve_load_data)
+            .collect();
+
+        entries.reverse();
+
+        let current_index = entries.len();
+
+        entries.push(current_load_data.clone());
+
+        let mut future_entries = self.joint_session_future(top_level_frame_id)
+            .map(&keep_load_data_if_top_frame)
+            .scan(current_load_data.clone(), &resolve_load_data)
+            .collect();
+
+        entries.append(&mut future_entries);
```

Comment:
Maybe. How do I append the current entry (using `chain(once())`?)  and also get the size of the past entries (to get the current index)?